test/unit/org/apache/cassandra/journal/IndexTest.java

```diff
@@ -0,0 +1,263 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.journal;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import org.junit.Test;
+
+import org.agrona.collections.IntHashSet;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.utils.Generators;
+import org.apache.cassandra.utils.TimeUUID;
+import org.quicktheories.core.Gen;
+import org.quicktheories.impl.Constraint;
+
+import static org.apache.cassandra.utils.TimeUUID.Generator.nextTimeUUID;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.quicktheories.QuickTheory.qt;
+
+public class IndexTest
+{
+    private static final int[] EMPTY = new int[0];
+
+    @Test
+    public void testInMemoryIndexBasics()
+    {
+        InMemoryIndex<TimeUUID> index = InMemoryIndex.create(TimeUUIDKeySupport.INSTANCE);
+
+        TimeUUID key0 = nextTimeUUID();
+        TimeUUID key1 = nextTimeUUID();
+        TimeUUID key2 = nextTimeUUID();
+        TimeUUID key3 = nextTimeUUID();
+        TimeUUID key4 = nextTimeUUID();
+
+        assertArrayEquals(EMPTY, index.lookUp(key0));
+        assertArrayEquals(EMPTY, index.lookUp(key1));
+        assertArrayEquals(EMPTY, index.lookUp(key2));
+        assertArrayEquals(EMPTY, index.lookUp(key3));
+        assertArrayEquals(EMPTY, index.lookUp(key4));
+
+        int val11 = 1100;
+        int val21 = 2100;
+        int val22 = 2200;
+        int val31 = 3100;
+        int val32 = 3200;
+        int val33 = 3300;
+
+        index.update(key1, val11);
+        index.update(key2, val21);
+        index.update(key2, val22);
+        index.update(key3, val31);
+        index.update(key3, val32);
+        index.update(key3, val33);
+
+        assertArrayEquals(EMPTY, index.lookUp(key0));
+        assertArrayEquals(new int[] { val11 }, index.lookUp(key1));
+        assertArrayEquals(new int[] { val21, val22 }, index.lookUp(key2));
+        assertArrayEquals(new int[] { val31, val32, val33 }, index.lookUp(key3));
+        assertArrayEquals(EMPTY, index.lookUp(key4));
+
+        assertEquals(key1, index.firstId());
+        assertEquals(key3, index.lastId());
+
+        assertFalse(index.mayContainId(key0));
+         assertTrue(index.mayContainId(key1));
+         assertTrue(index.mayContainId(key2));
+         assertTrue(index.mayContainId(key3));
+        assertFalse(index.mayContainId(key4));
+    }
+
+    @Test
+    public void testInMemoryIndexPersists() throws IOException
```

Comment:
this test is a more specific example of `prop`, ok to leave but is a duplicate now

Reply:
Will keep it for simple debugging for now, thanks.